analyzers/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,149 @@
+func readFileContent(filename string) (string, error) {
+	content, err := os.ReadFile(filename)
+	if err != nil {
+		return "", err
+	}
+
+	return string(content), nil
+}
+
+// getDiagnosticsFromFile uses the file content to return diagnostics with metadata like line number, content, etc.
+func getDiagnosticsFromFile(fileContent string, errors []validator.ErrorMeta) []Diagnostic {
+	diagnostics := []Diagnostic{}
+
+	lines := strings.Split(string(fileContent), "\n")
+
+	// Iterate over each error and check line-by-line.
+	for _, err := range errors {
+		for lineNum, line := range lines {
+			result := checkField(line, err.Field)
+			if result {
+				// If the line contains the field name, and if it doesn't have a comment prefix, then we can proceed to diagnostic generation.
+				// if strings.Contains(line, err.Field) && !strings.HasPrefix(line, "#") {
+				// Prepare code frame for the current line.
+				codeFrame := prepareCodeFrame(lineNum, lines)
+
+				// Generate a diagnostic.
+				diag := Diagnostic{
+					Line:         lineNum,
+					Codeframe:    codeFrame,
+					ErrorMessage: err.Message,
+				}
+
+				diagnostics = append(diagnostics, diag)
+			}
+		}
+	}
+
+	return diagnostics
+}
+
+// prepareCodeFrame prepares a code frame using the file content. The code frame is meant to be displayed on the console while reporting diagnostics.
+// NOTE: lineNum always starts from 0.
+func prepareCodeFrame(lineNum int, lines []string) string {
+	frame := ""
+
+	if lineNum <= 1 {
+		// Case 1: When the line is near the top of the file.
+		// Generate a frame with the current and next line only.
+		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum+1, lines[lineNum]))
+		frame += fmt.Sprintf("  %d | %s\n", lineNum+2, lines[lineNum+1])
+
+	} else if lineNum >= (len(lines) - 1) {
+		// Case 2: When the line is near the bottom of the file.
+		// Generate a frame with the current line only.
+		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum, lines[lineNum-1]))
+	} else {
+		// Case 3: When the line is in between the top and the bottom.
+		// Generate a frame with the the previous, current and the next line.
+		frame += fmt.Sprintf("  %d | %s\n", lineNum, lines[lineNum-1])
+		frame += aec.LightRedF.Apply(fmt.Sprintf("> %d | %s\n", lineNum+1, lines[lineNum]))
+		frame += fmt.Sprintf("  %d | %s\n", lineNum+2, lines[lineNum+1])
+	}
+
+	return frame
+}
+
+// checkField checks if the line contains the field.
+func checkField(line, field string) bool {
+	// We use fieldRegexp for checking if the current line is a field or not.
+	// If it's not a field, for example, comments for explanation, etc., then checkField should return false.
+	exp := regexp.MustCompile(fieldRegexp)
+
+	// Get index for the "field" named group.
+	matches := exp.FindStringSubmatch(line)
+	idx := exp.SubexpIndex("field")
+
+	// Capture match.
+	match := ""
+	if len(matches) != 0 && idx != -1 {
+		match = matches[idx]
+	}
+	match = strings.TrimSpace(match)
```

Comment:
What is `idx`? Can you please add a helpful comment for what is being done here?

Reply:
Resolved in https://github.com/deepsourcelabs/cli/pull/169/commits/61385c434b0e5fb9df909d816ff797e8101b16b8.

Reply:
Resolved in https://github.com/deepsourcelabs/cli/pull/169/commits/e9709d9e23dafe50fe102011140d1037f6435a6e.